libs/entitlement/src/main/java/org/elasticsearch/entitlement/initialization/EntitlementInitialization.java

```diff
@@ -328,6 +340,62 @@ private static PolicyManager createPolicyManager() {
         );
     }
 
+    private static Set<Path> pathSet(Path... paths) {
+        return Arrays.stream(paths).map(x -> x.toAbsolutePath().normalize()).collect(Collectors.toUnmodifiableSet());
+    }
+
+    // package visible for tests
+    static void validateFilesEntitlements(
+        Map<String, Policy> pluginPolicies,
+        PathLookup pathLookup,
+        Path configDir,
+        Path pluginsDir,
+        Path modulesDir,
+        Path libDir
+    ) {
+        var pluginReadAccessForbidden = pathSet(pluginsDir, modulesDir, libDir);
+        var pluginWriteAccessForbidden = pathSet(configDir);
+        for (var pluginPolicy : pluginPolicies.entrySet()) {
+            List<Path> readPaths = getFileDataStream(pluginPolicy.getValue()).flatMap(x -> x.resolvePaths(pathLookup)).toList();
+            List<Path> writePaths = getFileDataStream(pluginPolicy.getValue()).filter(x -> x.mode().equals(READ_WRITE))
+                .flatMap(x -> x.resolvePaths(pathLookup))
+                .toList();
+            validateLayerFilesEntitlements(pluginPolicy.getKey(), readPaths, pluginReadAccessForbidden, READ);
+            validateLayerFilesEntitlements(pluginPolicy.getKey(), writePaths, pluginWriteAccessForbidden, READ_WRITE);
+        }
+    }
+
+    private static Stream<FileData> getFileDataStream(Policy policy) {
+        return getFileDataStream(policy.scopes().stream().flatMap(x -> x.entitlements().stream()));
+    }
+
+    private static Stream<FileData> getFileDataStream(Stream<Entitlement> entitlements) {
+        return entitlements.filter(x -> x instanceof FilesEntitlement).flatMap(x -> ((FilesEntitlement) x).filesData().stream());
+    }
+
+    private static void validateLayerFilesEntitlements(
```

Comment:
What does `Layer` mean in this context? Is this anticipating renaming "component" to "layer"? If so, we might want to consider holding off on that until the renaming actually happens.

Reply:
We already have a mix naming here. I'm fine with keeping component for now though.